toolchain/check/check.cpp

```diff
@@ -279,29 +269,77 @@ auto CheckParseTrees(const SemIR::File& builtin_ir,
       continue;
     }
 
-    if (packaging && packaging->api_or_impl == Parse::Tree::ApiOrImpl::Impl) {
-      continue;
+    bool is_impl =
+        packaging && packaging->api_or_impl == Parse::Tree::ApiOrImpl::Impl;
+
+    // Add to the `api` map and diagnose duplicates. This occurs before the
+    // file extension check because we might emit both diagnostics in situation
+    // where the user forgets (or has syntax errors with) a package line
+    // multiple times.
+    if (!is_impl) {
+      auto [entry, success] = api_map.insert({import_key, &unit_info});
+      if (!success) {
+        llvm::StringRef prev_filename = entry->second->unit->tokens->filename();
+        if (packaging) {
+          CARBON_DIAGNOSTIC(DuplicateLibraryApi, Error,
+                            "Library's API previously provided by `{0}`.",
+                            llvm::StringRef);
+          unit_info.emitter.Emit(packaging->names.node, DuplicateLibraryApi,
+                                 prev_filename);
+        } else {
+          CARBON_DIAGNOSTIC(DuplicateMainApi, Error,
+                            "Main//default previously provided by `{0}`.",
+                            llvm::StringRef);
+          // Use the invalid node because there's no node to associate with.
+          unit_info.emitter.Emit(Parse::Node::Invalid, DuplicateMainApi,
+                                 prev_filename);
+        }
+      }
     }
 
-    auto [entry, success] = api_map.insert({import_key, &unit_info});
-    if (!success) {
-      llvm::StringRef prev_filename = entry->second->unit->tokens->filename();
-      if (packaging) {
-        CARBON_DIAGNOSTIC(DuplicateLibraryApi, Error,
-                          "Library's API previously provided by `{0}`.",
-                          llvm::StringRef);
-        unit_info.emitter.Emit(packaging->names.node, DuplicateLibraryApi,
-                               prev_filename);
-      } else {
-        CARBON_DIAGNOSTIC(DuplicateMainApi, Error,
-                          "Main//default previously provided by `{0}`.",
-                          llvm::StringRef);
-        // Use the invalid node because there's no node to associate with.
-        unit_info.emitter.Emit(Parse::Node::Invalid, DuplicateMainApi,
-                               prev_filename);
+    // Validate file extensions. Note imports rely the packaging directive, not
+    // the extension.
+    auto filename = unit_info.unit->tokens->filename();
+    static constexpr llvm::StringLiteral ApiExt = ".carbon";
+    static constexpr llvm::StringLiteral ImplExt = ".impl.carbon";
+    bool is_api_with_impl_ext = !is_impl && filename.ends_with(ImplExt);
+    auto want_ext = is_impl ? ImplExt : ApiExt;
+    if (is_api_with_impl_ext || !filename.ends_with(want_ext)) {
+      CARBON_DIAGNOSTIC(IncorrectExtension, Error,
+                        "File extension of `{0}` required for `{1}`.",
+                        llvm::StringLiteral, Lex::TokenKind);
```

Comment:
What do you think about splitting the diagonstic approach up a bit?

It would seem a bit cleaner to me to have a diagnostic for any file that doesn't end in `.carbon`, without mentioning anything about `api` or `impl` at all.

And then a separate diagnostic, located at the actual `impl` keyword, that complains about the extension not being `.impl.carbon` specifically.

And then similar to what you have below, but located at the `api` keyword (or the package directive eventually if we remove that), and explains that this *isn't* an `impl` and so can't have the extension `.impl.carbon`.

WDYT?

Happy for this to be a follow-up though, this PR is still an improvement as now we're diagnosing things. Just thinking about how to make the errors be best understood.

Reply:
To be clear, I'm interpreting your suggestion as:

```
foo.x: ERROR: Incorrect file extension.
foo.x: File extension of `.carbon` required for `api`.
package Foo api;
^
```

```
bar.impl.carbon: ERROR: Incorrect file extension.
bar.impl.carbon: File extension of `.carbon` required for `api`.
package Bar api;
^
bar.impl.carbon: File extension of `.impl.carbon` only allowed for `impl`.
```

I think the difference here is an error declaring the classification of issue, versus an error trying to be informative. It seemed like rust may have gone in that direction (e.g., `error[E0308]: mismatched types`, versus a message including the types). But I think if you look at errors elsewhere in the toolchain, that's a different direction than we've headed: I would advocate at least move uniformly in errors.